src/states/node.rs

```diff
@@ -809,6 +829,150 @@ impl Node {
+                                        result);
+        self.add_to_routing_table(&pub_id, &peer_id).extract(&mut result);
+        if let Some(prefix) = self.peer_mgr.routing_table().find_group_prefix(&candidate_name) {
+            self.send_section_list_signature(prefix, None);
+            if prefix == *self.peer_mgr.routing_table().our_prefix() {
+                // if the node joined our section, send signatures for all section lists to it
+                for pfx in self.peer_mgr.routing_table().prefixes() {
+                    self.send_section_list_signature(pfx, Some(candidate_name));
+                }
+            }
+        }
+        result.map(Ok)
+    }
+
+    fn handle_node_approval(&mut self,
+                            groups: Vec<(Prefix<XorName>, Vec<PublicId>)>)
+                            -> Evented<Result<(), RoutingError>> {
+        let mut events = Evented::empty();
+        if self.is_approved {
+            warn!("{:?} Received duplicate NodeApproval.", self);
+            return events.with_value(Ok(()));
+        }
+
+        self.peer_mgr.populate_routing_table(&groups);
+
+        // TODO: is this neccssary as this node is not approved as a full node by the section yet
+        let our_prefix = *self.peer_mgr.routing_table().our_prefix();
+        self.send_section_list_signature(our_prefix, None);
+
+        let name = *self.name();
+        if let Err(error) = self.send_routing_message(RoutingMessage {
+                src: Authority::ManagedNode(name),
+                dst: Authority::Section(name),
+                content: MessageContent::ApprovalConfirmation,
+            })
+            .extract(&mut events) {
+            debug!("{:?} Failed sending ApprovalConfirmation: {:?}", self, error);
+        }
+
+        trace!("{:?} received {:?} on NodeApproval.", self, groups);
+
+        for group in &groups {
+            for pub_id in &group.1 {
+                if !self.peer_mgr.routing_table().has(pub_id.name()) {
+                    debug!("{:?} Sending connection info to {:?} on NodeApproval.",
+                           self,
+                           pub_id);
+                    let src = Authority::ManagedNode(*self.name());
+                    let node_auth = Authority::ManagedNode(*pub_id.name());
+                    if let Err(error) = self.send_connection_info_request(*pub_id, src, node_auth)
+                        .extract(&mut events) {
+                        debug!("{:?} - Failed to send connection info to {:?}: {:?}",
+                               self,
+                               pub_id,
+                               error);
+                    }
+                }
+            }
+        }
+
+        events.add_event(Event::Connected);
+        for name in self.peer_mgr.routing_table().iter() {
+            events.add_event(Event::NodeAdded(*name, self.peer_mgr.routing_table().clone()));
```

Comment:
I wonder whether we should even send this. Maybe add a `TODO` to try out removing that. I don't think safe_vault relies on `NodeAdded` events in the joining node.